pkg/store/prometheus.go

```diff
@@ -290,6 +290,32 @@ func labelsMatches(lset labels.Labels, ms []storepb.LabelMatcher) (bool, []store
 	return true, newMatcher, nil
 }
 
+func labelsMatchesAndTranslate(lset labels.Labels, ms []storepb.LabelMatcher) (bool, []labels.Matcher, error) {
+	tms, err := translateMatchers(ms)
+	if err != nil {
+		return false, nil, err
+	}
+
+	if len(lset) == 0 {
+		return true, tms, nil
+	}
+
+	for i := len(tms) - 1; i >= 0; i-- {
+		tm := tms[i]
+		v := lset.Get(tm.Name())
+		if v == "" {
+			continue
+		}
+
+		if !tm.Matches(v) {
+			return false, nil, nil
+		}
+
+		tms = append(tms[:i], tms[i+1:]...)
```

Comment:
I have a question here. Why do we remove the match label from request `LabelMatcher`? I follow the implemete of the `labelsMatches` method above.

Reply:
Sorry what do you mean remove? You mean `continue` part on line 315? This feels wrong. It's bit different to what we have here: https://github.com/thanos-io/thanos/blob/d02488d9be253223393515a0f9c1fce5318ca3af/pkg/store/prometheus.go#L311

That's why.. can we have just one function for matching + selector labels? (:

Reply:
Re-reading the code, I'm sorry that I misunderstood the `external_labels` that those should not be in the `storepb.LabelMatcher`, because those would be added to any series. Here the `selector-labels` are as same as the `external_labels` used by store gateway sharding selector. We assume that all of series in the bucket queried by `store` component have the `selector-labels` pairs.

I have reused the `matchesExternalLabels` function to check `selector-labels`, and do `translateMatcher` again to get the new `labels.Matcher`.